src/transformers/generation/logits_process.py

```diff
@@ -1102,3 +1102,82 @@ def __call__(self, input_ids, scores):
+    weight `rescale_factor` the conditional ones to smooth the effect and increase output quality.
+
+    See [the paper](https://arxiv.org/abs/2306.17806) for more information.
+
+    Args:
+        guidance_scale (float):
+            The guidance scale for classifier free guidance (CFG). CFG is enabled by setting `guidance_scale > 1`.
+            Higher guidance scale encourages the model to generate samples that are more closely linked to the input
+            prompt, usually at the expense of poorer quality.
+        unconditional_ids (`torch.LongTensor` of shape `(batch_size, sequence_length)`):
+            Indices of input sequence tokens in the vocabulary for the unconditional branch.
+        model:
+            The LM computing the unconditional scores. Supposedly the same as the one computing the conditional scores.
+            Both models must use the same tokenizer.
+        smooth_factor (float):
+            The interpolation weight for CFG Rescale. 1 means no rescaling, 0 reduces to the conditional scores without
+            CFG. Turn it lower if the output degenerates.
+
+    Examples:
+
+    ```python
+    >>> # base prompt emphasis example
+    >>> from transformers import AutoTokenizer, AutoModelForCausalLM
+
+    >>> model = AutoModelForCausalLM.from_pretrained("gpt2")
+    >>> tokenizer = AutoTokenizer.from_pretrained("gpt2")
+    >>> inputs = tokenizer(["Today, a dragon flew over Paris, France,"], return_tensors="pt")
+    >>> summary_ids = model.generate(inputs["input_ids"], guidance_scale=1.5)
+    >>> print(tokenizer.batch_decode(summary_ids, skip_special_tokens=True)[0])
+    The dragon flew over Paris, France, landing in Lyon, a city of a few million. Dragon-flying was a new form of
+    transport, and the dragon was the first in Europe.
+
+    >>> # with a negative prompt
+    >>> neg_inputs = tokenizer(["A very happy event happened,"], return_tensors="pt")
+    >>> summary_ids = model.generate(
+    ...     inputs["input_ids"], guidance_scale=5, cfg_rescale=0.3, negative_prompt=neg_inputs["input_ids"]
+    ... )
+    >>> print(tokenizer.batch_decode(summary_ids, skip_special_tokens=True)[0])
+    The dragon flew over Paris, France, crashing into Notre Dame Cathedral in the French capital killing at least 127
+    people and injuring more than 350.
+    ```
+    """
+
+    def __init__(self, guidance_scale, unconditional_ids, model, rescale_factor=1.0):
+        self.guidance_scale = guidance_scale
+        self.unconditional_ids = unconditional_ids
+        self.model = model
+        self.out = None
+        self.rescale_factor = rescale_factor
+
+    def __call__(self, input_ids, scores):
+        scores = torch.nn.functional.log_softmax(scores, dim=-1)
+        if self.guidance_scale == 1:
+            return scores
+
+        if self.out is None:
+            self.out = self.model(
+                self.unconditional_ids, attention_mask=torch.ones_like(self.unconditional_ids[:, -1:]), use_cache=True
```

Comment:
I hope this isn't a dumb question, but it looks like the `attention_mask` will only look like `tensor([[1]])` but your `unconditional_ids` can have a lot more than one token, right?  E.g. if someone passed a set of unconditional ids associated with a `negative_prompt` as in your example.

I don't know for certain how `attention_mask` expands if it has fewer entries than the ids passed in, but if I'm reading `create_extended_attention_mask_for_decoder` right, it looks like it expands to all of the inputs, by filling with `1`s.  If that's the case, why create an `attention_mask` at all, since it _should_ result in no change in the end?

Reply:
You're totally right. Broadcasting made it work silently I assume.

Good eyes. Though this does not need fixing since this attention_mask is meant to change but I'm waiting for the maintainers' guidelines on this.